source/common/http/http2/codec_impl.cc

```diff
@@ -283,17 +291,64 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
       stream->headers_->addViaMove(std::move(key), std::move(stream->cookies_));
     }
 
-    if (frame->headers.cat == NGHTTP2_HCAT_REQUEST || frame->headers.cat == NGHTTP2_HCAT_RESPONSE) {
+    if (frame->headers.cat == NGHTTP2_HCAT_REQUEST && stream->headers_->Expect() &&
+        0 == StringUtil::caseInsensitiveCompare(stream->headers_->Expect()->value().c_str(),
+                                                Headers::get().ExpectValues._100Continue.c_str())) {
+      // Deal with expect: 100-continue here since higher layers are never going to do anything
+      // other than say to continue so that we can respond before request complete if necessary.
+      std::vector<nghttp2_nv> final_headers;
+      StreamImpl::buildHeaders(final_headers, *CONTINUE_HEADER);
+      int rc = nghttp2_submit_headers(session_, 0, stream->stream_id_, nullptr, &final_headers[0],
+                                      final_headers.size(), nullptr);
+      ASSERT(rc == 0);
```

Comment:
Based on logic flow, this function should never fail, so I made it a debug only assert. We do this in a few other places in this file.